File: java/leetcode/src/main/java/com/leetcode/graph/IsGraphBipartite_785.java
package com.leetcode.graph;

public class IsGraphBipartite_785 {
    private boolean isBipartite = true;

    private int nextColor(int color) {
        if (color == 2) return 1;
        return 2;
    }

    public boolean isBipartite(int[][] graph) {
        return dfs_solution(graph);
    }

    private boolean dfs_solution(int[][] graph) {
        int n = graph.length;
        int[] visited = new int[n];

        for(int i = 0; i < n; i++) {
            if (visited[i] == 0) {
                dfs(graph, i, 1, visited);
            }
            if (!isBipartite) return isBipartite;
        }

        return isBipartite;
    }

    private void dfs(int[][] graph, int s, int color, int[] visited) {
        visited[s]  = color;
        int nextColor = nextColor(color);
        for (int v : graph[s]) {

            if (!isBipartite) return;

            if (visited[v] == 0) {
                dfs(graph, v, nextColor, visited);
            } else {
                if (color == visited[v]) {
                    isBipartite = false;
                }
            }
        }
    }
}

File: js/algrithm/Graph/Is_Graph_Bipartite_785.js
/**
 * @param {number[][]} graph
 * @return {boolean}
 */
var isBipartite = function(graph) {
    //const res = dfs_solution(graph)
    const res = bfs_solution(graph)
    return res
};

function dfs_solution(graph) {
    const n = graph.length
    const visited = Array(n).fill(0)
    for (let i = 0; i < n; i++) {
        if (visited[i] === 0) {
            if (!dfs(graph, i, visited, 1)) return false
        }
    }
    return true
}

function dfs(graph, i, visited, color) {
    visited[i] = color
    for (let next of graph[i]) {
        if (visited[next] === color) return false
        if (visited[next] === 0) {
            if (!dfs(graph, next, visited, -1 * color)) return false
        }
    }
    return true
}

function bfs_solution(graph) {
    const n = graph.length
    const visited = Array(n).fill(0)
    
    for (let i = 0; i < n; i++) {
        const queue = []
        if (visited[i] === 0) {
            queue.push([i, 1])
            visited[i]= 1
        }

        while (queue.length) {
            const [cur, color] = queue.shift()
            for (let next of graph[cur]) {
                if (visited[next] === color) return false
                if (visited[next] === 0) {
                    queue.push([next, -1 * color])
                    visited[next] = -1 * color
                }
            }
        }

    }

    return true
}